### DISCUSSION TITLE
how do you enable HMR with vite-node?

### DISCUSSION THREAD

**Discussion by dionsaur84** [vitest-dev/vitest#2588] [Q&A]
I can't figure out how to enable HMR using vite-node and express.

For example, here is my `dev` script inside of `package.json`: 

```json
"scripts": {
    "dev": "vite build --watch && vite-node ./src/entries/server.js --watch"
},
```

This works great, it builds my frontend code & listens on my backend code. It also watches for changes and rebuilds/restarts the server when files change.

However HMR is not working at all.

Here is my `vite.config.js` file:

```javascript
import reactPlugin from '@vitejs/plugin-react'
import { defineConfig } from 'vite'

export default defineConfig({
    build: {
        rollupOptions: {
            input: './src/entries/client.jsx',
            output: {
                entryFileNames: '[name].js',
                assetFileNames: '[name].[ext]',
            },
        },
        outDir: './public/dist',
    },
    publicDir: false,
    plugins: [
        reactPlugin(),
    ],
})
```

The `./src/entries/server.js` file just starts an express server:

```javascript
app.listen(env.APP_PORT, function () {
    console.log(`app listening on http://localhost:${env.APP_PORT}`)
})
```

Is there a way to make HMR work with this type of setup?

**Comment by meesvandongen** [comment]
From what I understand, HMR is on by default via the vite server. The watch mode is another feature entirely for restarting on changes.
